### ISSUE TITLE
Docs on how to contribute to Xonsh

### ISSUE THREAD

**Issue by santagada** [xonsh/xonsh#1430]
specially talk about how to setup a dev environment with amalgamation deactivated, I tried this steps and failed:
- created a vox env
- installed requirements-test
- python3 setup.py develop
- py.test

it resulted in this error:

```
 $  py.test
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 3.5.2, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /Users/santagada/devel/xonsh, inifile:
collected 1131 items

tests/test_aliases.py .........
tests/test_ast.py ...
tests/test_builtins.py s.....................
tests/test_contexts.py .........................
tests/test_dirstack.py ....
tests/test_environ.py ..............................s
tests/test_execer.py ss..................
tests/test_foreign_shells.py .....s
tests/test_history.py ......
tests/test_imphooks.py ....
tests/test_inspectors.py .
tests/test_lazyasd.py .
tests/test_lazyjson.py ...................
tests/test_lexer.py .....................................
tests/test_main.py .................
tests/test_man.py .
tests/test_news.py .................
tests/test_parser.py .....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
tests/test_ptk_history.py .
tests/test_ptk_multiline.py .......
tests/test_replay.py sss
tests/test_tools.py ................................................................................................................................................................................................................................................................................................ssss...s
tests/test_wizard.py ...
tests/test_xontribs.py .

================================================================================================== 1118 passed, 13 skipped in 7.14 seconds ===================================================================================================
Exception ignored in: <bound method Execer.__del__ of <xonsh.__amalgam__.Execer object at 0x10ed742b0>>
Traceback (most recent call last):
  File "/Users/santagada/devel/xonsh/xonsh/__amalgam__.py", line 16113, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <generator object get_tokens at 0x10f29b9e8>
Traceback (most recent call last):
  File "/Users/santagada/devel/xonsh/xonsh/__amalgam__.py", line 9931, in get_tokens
TypeError: catching classes that do not inherit from BaseException is not allowed
```

**Comment by gforsyth** [comment]
Thanks for reporting, @santagada!  We can definitely use some more documentation on this.  
If you want to disable amalgamation for testing/development, you can do (from within `xonsh`)

```console
$XONSH_DEBUG=1
xonsh
```

or from Bash/zsh/whatever
```console
env XONSH_DEBUG=1 xonsh
```

Any help with updating contributor documentation is greatly appreciated!

**Comment by bobhy** [comment]
I found this problem trying to debug in Windows, and eventually the same fix.   Xonsh is a bit tricky to set up for debug compared to other, even much bigger, projects like Django. I think there are two key tricks to master: to disable amalgamation or to set breakpoints in amalgamated code;  and setting the debugger to to invoke Xonsh with a specific command line to hit your breakpoints, rather than trying to run the  console under the debugger (still a problem for me with PyCharm)  

I'll take a crack at some language in the dev guide if some of the seniors on the project would provide some pointers on how they do debugging and why amalgamation.

**Comment by scopatz** [comment]
> Xonsh is a bit tricky to set up for debug compared to other, even much bigger, projects like Django.

Well we are much younger and much smaller than Django so I think we still have some work on our debugging story :)

> why amalgamation.

Amalgamation is there to help ensure fast startup times while maintaining fast run times and idiomatic Python programming.

**Comment by scopatz** [comment]
Also, please see the amalgamate readme: https://github.com/xonsh/amalgamate#amalgamate

**Comment by gforsyth** [comment]
@bobhy re: debugging, i use `pdb` or `ipdb`.  `ipdb` sometimes has trouble for reasons I don't entirely understand.  But I've found that the best approach for me to is to put in my `set_trace` wherever I want it and then to run `xonsh` from within a Python interpreter, a la

```python
from xonsh.shell import Shell
s = Shell()
s.default('ls')
```

if you set `$XONSH_DEBUG=1` in your environment before starting the Python interpreter that should get passed through when you start the `Shell` instance within Python and disable amalgamation.

**Comment by dotsdl** [comment]
Added to #3225

**Comment by anki-code** [comment]
After 5 years this issue should be actualized or create new issue. Closed until this.
